### ISSUE TITLE
The recorded step number in profiler is wrong.

### ISSUE THREAD

**Issue by fwenguang** [pytorch/pytorch#121126]
### 🐛 Describe the bug

``` python
def test():
    import torch
    x = torch.randn(10, 10).cuda()
    y = torch.randn(10, 10).cuda()
    with torch.profiler.profile(
            activities=[
                torch.profiler.ProfilerActivity.CPU,
                torch.profiler.ProfilerActivity.CUDA,
            ],
            schedule=torch.profiler.schedule(
                wait=1,
                warmup=2,
                active=3
            ),
            on_trace_ready=torch.profiler.tensorboard_trace_handler('./log'),
        ) as p:
        for _ in range(6):
           z = torch.matmul(x, y)
           p.step()

test()
```
After running above code. I found the first recorded step is Step#2. 
![image](https://github.com/pytorch/pytorch/assets/95677191/3a4e815e-1475-489f-a74a-bf5400a6d0d3)

According to the scheduling strategy, profiling should wait 1 step and warmup 2 steps, so the first recorded step should be Step#3.

`p.step()` [Signals the profiler that the next profiling step has started.](https://github.com/pytorch/pytorch/blob/main/torch/profiler/profiler.py#L652), so the recorded step number should be next step's.

### Versions

main

cc @robieta @chaekit @aaronenyeshi @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb @dzhulgakov @davidberard98

**Comment by fwenguang** [comment]
Fixed by https://github.com/pytorch/pytorch/pull/121127, closing this issue.

**Comment by Qizhi697** [comment]
@fwenguang 
When wait=1, the profiler recorded fn[dispatch/conbine] 30/10=3 steps.

```python
schedule = torch.profiler.schedule(wait=1, warmup=1, active=2, repeat=0)
with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU,torch.profiler.ProfilerActivity.CUDA], schedule=schedule) as prof:
      for i in range(4):
          dist.all_reduce(torch.ones(1, dtype=torch.float, device='cuda'))
          for _ in range(10):
              fn()
          prof.step()
```

```table
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                                                                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                                                       ProfilerStep*         0.00%       0.000us         0.00%       0.000us       0.000us     755.137ms        63.45%     755.137ms     188.784ms             4  
void deep_ep::internode_ll::dispatch<3, 10, 7168>(void*, float*, int*, long*, void*, int*, void*,...         0.00%       0.000us         0.00%       0.000us       0.000us     723.111ms        60.76%     723.111ms      24.104ms            30  
void deep_ep::internode_ll::combine<3, 10, 7168, 9>(void*, void*, int*, void*, void const*, long ...         0.00%       0.000us         0.00%       0.000us       0.000us     408.413ms        34.31%     408.413ms      13.614ms            30  
                                                                                       ProfilerStep*         0.09%       1.129ms         0.20%       2.360ms       1.180ms       0.000us         0.00%      58.665ms      29.333ms             2  
                                                                                    c10d::allreduce_         0.01%      65.902us         0.02%     239.234us     119.617us       0.000us         0.00%      58.662ms      29.331ms             2  
                                                                                  record_param_comms         0.01%     137.480us         0.02%     180.768us      45.192us      58.662ms         4.93%      58.662ms      14.666ms             4  
                     ncclDevKernel_AllReduce_Sum_f32_TREE_LL(ncclDevComm*, unsigned long, ncclWork*)         0.00%       0.000us         0.00%       0.000us       0.000us      58.662ms         4.93%      58.662ms      29.331ms             2  
                                                                                     nccl:all_reduce         0.00%       0.000us         0.00%       0.000us       0.000us      58.662ms         4.93%      58.662ms      29.331ms             2  
                                                                                          aten::ones         0.00%      20.084us         0.01%     121.210us      60.605us       0.000us         0.00%       2.656us       1.328us             2  
                                                                                         aten::fill_         0.00%      25.991us         0.01%      59.970us      29.985us       2.656us         0.00%       2.656us       1.328us             2  
void at::native::vectorized_elementwise_kernel<4, at::native::FillFunctor<float>, std::array<char...         0.00%       0.000us         0.00%       0.000us       0.000us       2.656us         0.00%       2.656us       1.328us             2  
                                                                                         aten::empty         0.04%     458.314us         0.04%     518.716us       5.085us       0.000us         0.00%       0.000us       0.000us           102  
                                                                                      cudaEventQuery         0.01%      88.335us         0.01%      88.335us       0.736us       0.000us         0.00%       0.000us       0.000us           120  
                                                                                    cudaLaunchKernel         0.00%      33.979us         0.00%      33.979us      16.989us       0.000us         0.00%       0.000us       0.000us             2  
                                                                               cudaStreamIsCapturing         0.00%       2.371us         0.00%       2.371us       1.185us       0.000us         0.00%       0.000us       0.000us             2  
                                                                                     cudaEventRecord         0.01%      69.580us         0.01%      69.580us       0.756us       0.000us         0.00%       0.000us       0.000us            92  
                                                                                 cudaStreamWaitEvent         0.01%      73.832us         0.01%      73.832us       0.839us       0.000us         0.00%       0.000us       0.000us            88  
                                                                                     nccl:all_reduce         0.00%       0.000us             0     128.812us      64.406us       0.000us         0.00%       0.000us       0.000us             2  
                                                                         cudaStreamGetCaptureInfo_v2         0.00%       2.234us         0.00%       2.234us       1.117us       0.000us         0.00%       0.000us       0.000us             2  
                                                                                 cudaLaunchKernelExC         0.02%     211.872us         0.02%     211.872us       5.045us       0.000us         0.00%       0.000us       0.000us            42  
                                                                            cudaPointerGetAttributes         0.00%      14.738us         0.00%      14.738us       0.737us       0.000us         0.00%       0.000us       0.000us            20  
                                                                                     aten::transpose         0.00%      42.192us         0.00%      54.164us       2.708us       0.000us         0.00%       0.000us       0.000us            20  
                                                                                    aten::as_strided         0.00%      11.972us         0.00%      11.972us       0.599us       0.000us         0.00%       0.000us       0.000us            20  
                                                                               cudaDeviceSynchronize        99.80%        1.196s        99.80%        1.196s        1.196s       0.000us         0.00%       0.000us       0.000us             1  
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------ 
```

When change wait=0, the profiler only recorded fn[dispatch/conbine] 10/10=1 step.

```python
schedule = torch.profiler.schedule(wait=0, warmup=1, active=2, repeat=0)
with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU,torch.profiler.ProfilerActivity.CUDA], schedule=schedule) as prof:
      for i in range(4):
          dist.all_reduce(torch.ones(1, dtype=torch.float, device='cuda'))
          for _ in range(10):
              fn()
          prof.step()
```

```table
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                                                                Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
void deep_ep::internode_ll::dispatch<3, 10, 7168>(void*, float*, int*, long*, void*, int*, void*,...         0.00%       0.000us         0.00%       0.000us       0.000us     368.628ms        76.52%     368.628ms      36.863ms            10  
void deep_ep::internode_ll::combine<3, 10, 7168, 9>(void*, void*, int*, void*, void const*, long ...         0.00%       0.000us         0.00%       0.000us       0.000us     113.087ms        23.48%     113.087ms      11.309ms            10  
                                                                                       ProfilerStep*         0.01%      44.027us         0.01%      44.027us      44.027us       0.000us         0.00%       0.000us       0.000us             1  
                                                                               cudaDeviceSynchronize        99.99%     497.752ms        99.99%     497.759ms     497.759ms       0.000us         0.00%       0.000us       0.000us             1  
                                                                                      cudaEventQuery         0.00%       7.215us         0.00%       7.215us       3.607us       0.000us         0.00%       0.000us       0.000us             2  
----------------------------------------------------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------ 
```

**Comment by Qizhi697** [comment]
[https://github.com/pytorch/pytorch/issues/148314](url)
